jupyter_server/services/kernels/v3/client.py

```diff
@@ -0,0 +1,733 @@
+import asyncio
+import time
+import typing as t
+from datetime import datetime, timezone
+
+from jupyter_client.asynchronous.client import AsyncKernelClient
+from jupyter_client.channels import AsyncZMQSocketChannel
+from jupyter_client.channelsabc import ChannelABC
+from traitlets import HasTraits, Type
+
+from .message_utils import encode_channel_in_message_dict, parse_msg_id
+from .states import ExecutionStates
+
+
+class NamedAsyncZMQSocketChannel(AsyncZMQSocketChannel):
+    """Prepends the channel name to all message IDs to this socket."""
+
+    channel_name = "unknown"
+
+    def send(self, msg):
+        """Send a message with automatic channel encoding."""
+        msg = encode_channel_in_message_dict(msg, self.channel_name)
+        return super().send(msg)
+
+
+class ShellChannel(NamedAsyncZMQSocketChannel):
+    """Shell channel that automatically encodes 'shell' in outgoing msg_ids."""
+
+    channel_name = "shell"
+
+
+class ControlChannel(AsyncZMQSocketChannel):
```

Comment:
Should this extend from NamedAsyncZMQSocketChannel if encoding channel name is desired?